core/matrix/sellp.hpp

```diff
@@ -0,0 +1,327 @@
+                       size_type idx) noexcept
+    {
+        return this->get_col_idxs()[this->linearize_index(row, slice_set, idx)];
+    }
+
+    /**
+     * @copydoc Sellp::col_at(size_type, size_type, size_type)
+     */
+    index_type col_at(size_type row, size_type slice_set, size_type idx) const
+        noexcept
+    {
+        return this
+            ->get_const_col_idxs()[this->linearize_index(row, slice_set, idx)];
+    }
+
+protected:
+    /**
+     * Creates an uninitialized Sellp matrix of the specified size.
+     *    (The total_cols is set to be the number of slice times the number
+     *     of cols of the matrix.)
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     */
+    Sellp(std::shared_ptr<const Executor> exec, const dim &size = dim{})
+        : Sellp(std::move(exec), size,
+                ceildiv(size.num_rows, default_slice_size) * size.num_cols)
+    {}
+
+    /**
+     * Creates an uninitialized Sellp matrix of the specified size.
+     *    (The slice_size and stride_factor are set to the default values.)
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param total_cols   number of the sum of all cols in every slice.
+     */
+    Sellp(std::shared_ptr<const Executor> exec, const dim &size,
+          size_type total_cols)
+        : Sellp(std::move(exec), size, default_slice_size,
+                default_stride_factor, total_cols)
+    {}
+
+    /**
+     * Creates an uninitialized Sellp matrix of the specified size.
+     *
+     * @param exec  Executor associated to the matrix
+     * @param size  size of the matrix
+     * @param slice_size  number of rows in each slice
+     * @param stride_factor  factor for the stride in each slice (strides
+     *                        should be multiples of the stride_factor)
+     * @param total_cols   number of the sum of all cols in every slice.
+     */
+    Sellp(std::shared_ptr<const Executor> exec, const dim &size,
+          size_type slice_size, size_type stride_factor, size_type total_cols)
+        : EnableLinOp<Sellp>(exec, size),
+          values_(exec, slice_size * total_cols),
+          col_idxs_(exec, slice_size * total_cols),
+          slice_lengths_(exec, (size.num_rows == 0)
+                                   ? 0
+                                   : ceildiv(size.num_rows, slice_size)),
+          slice_sets_(exec, (size.num_rows == 0)
+                                ? 0
+                                : ceildiv(size.num_rows, slice_size) + 1),
+          slice_size_(slice_size),
+          stride_factor_(stride_factor),
+          total_cols_(total_cols)
+    {}
+
+    void apply_impl(const LinOp *b, LinOp *x) const override;
+
+    void apply_impl(const LinOp *alpha, const LinOp *b, const LinOp *beta,
+                    LinOp *x) const override;
+
+    size_type linearize_index(size_type row, size_type slice_set,
```

Comment:
Maybe this function could be inline, or does C++ do that automatically? Doing so could improve performance. of the functions `val_at`, `col_at`, etc.

Reply:
C++ does it automatically. In the following code:

```c++
class foo {
    int f() { return 5; }
    int g();
};

int foo::g() { return 5; }
```

`f` is `inline`, `g` is not. ([godbolt](https://godbolt.org/g/PQUdps))

Reply:
You are correct. [Here](https://godbolt.org/g/Lrd3yP) is a better example IMHO. You need at least -O1 for it to appear (sounds normal).

Reply:
Relevant passage of the standard proving that it's indeed specified:
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf section 7.1.2 paragraph 3.

Reply:
The thing is: `inline` in C++ does not mean "the compiler will not generate a function call".  So even without -O1 the function is still inline, but the compiler just calls it as any other normal function. It is a "hint" that the function should be inline though.  More about it in [__\[dcl.inline\]__](http://eel.is/c++draft/dcl.inline) and [__\[class.mfct\]__](http://eel.is/c++draft/class.mfct).

That's why I didn't do your example, as it gets confusing, especially because with -O1 even non-inline functions get inlined by the compiler: https://godbolt.org/g/zNZVAU

Reply:
I did not catch this comment. So is inline necessary in this case (val_at, col_at)?

Reply:
In the end it is not. After @gflegar 's explanation and looking through it, it is not necessary. It is implied by the standard.

Reply:
Got it. Thank you for the explanation.